dbms/src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -26,69 +27,156 @@ namespace DB
+
+            throw Exception(error_message.str(), ErrorCodes::TOO_MANY_REDIRECTS);
+        }
+    }
+
+    virtual ~UpdatableSessionBase()
+    {
+    }
+};
+
+
 namespace detail
 {
-    template <typename SessionPtr>
+    template <typename UpdatableSessionPtr>
     class ReadWriteBufferFromHTTPBase : public ReadBuffer
     {
     protected:
         Poco::URI uri;
         std::string method;
 
-        SessionPtr session;
+        UpdatableSessionPtr session;
         std::istream * istr; /// owned by session
         std::unique_ptr<ReadBuffer> impl;
+        std::function<void(std::ostream &)> out_stream_callback;
+        const Poco::Net::HTTPBasicCredentials & credentials;
         std::vector<Poco::Net::HTTPCookie> cookies;
 
-    public:
-        using OutStreamCallback = std::function<void(std::ostream &)>;
-
-        explicit ReadWriteBufferFromHTTPBase(
-            SessionPtr session_,
-            Poco::URI uri_,
-            const std::string & method_ = {},
-            OutStreamCallback out_stream_callback = {},
-            const Poco::Net::HTTPBasicCredentials & credentials = {},
-            size_t buffer_size_ = DBMS_DEFAULT_BUFFER_SIZE)
-            : ReadBuffer(nullptr, 0)
-            , uri{uri_}
-            , method{!method_.empty() ? method_ : out_stream_callback ? Poco::Net::HTTPRequest::HTTP_POST : Poco::Net::HTTPRequest::HTTP_GET}
-            , session{session_}
+        std::istream * call(const Poco::URI uri_, Poco::Net::HTTPResponse & response)
         {
             // With empty path poco will send "POST  HTTP/1.1" its bug.
             if (uri.getPath().empty())
                 uri.setPath("/");
 
-            Poco::Net::HTTPRequest request(method, uri.getPathAndQuery(), Poco::Net::HTTPRequest::HTTP_1_1);
-            request.setHost(uri.getHost()); // use original, not resolved host name in header
+            Poco::Net::HTTPRequest request(method, uri_.getPathAndQuery(), Poco::Net::HTTPRequest::HTTP_1_1);
+            request.setHost(uri_.getHost()); // use original, not resolved host name in header
 
             if (out_stream_callback)
                 request.setChunkedTransferEncoding(true);
 
             if (!credentials.getUsername().empty())
                 credentials.authenticate(request);
 
-            Poco::Net::HTTPResponse response;
-
             LOG_TRACE((&Logger::get("ReadWriteBufferFromHTTP")), "Sending request to " << uri.toString());
 
+            auto sess = session->getSession();
+
             try
             {
-                auto & stream_out = session->sendRequest(request);
+                auto & stream_out = sess->sendRequest(request);
 
                 if (out_stream_callback)
                     out_stream_callback(stream_out);
 
-                istr = receiveResponse(*session, request, response);
+                istr = receiveResponse(*sess, request, response, true);
```

Comment:
That's suspicious, need double check it...

Reply:
receiveResponse before the change only accepted 200 code responses. Last parameter accepts redirects codes (307,...) as ok. That loop accepts redirects and checks max_redirects limit.